program-runtime/src/compute_budget.rs

```diff
@@ -102,35 +102,54 @@ impl ComputeBudget {
         &mut self,
         message: &SanitizedMessage,
         requestable_heap_size: bool,
+        default_units_per_instruction: bool,
     ) -> Result<u64, TransactionError> {
+        let mut num_instructions = message.instructions().len();
         let mut requested_additional_fee = 0;
+        let mut requested_units = None;
+
         let error = TransactionError::InstructionError(0, InstructionError::InvalidInstructionData);
-        // Compute budget instruction must be in the 1st 3 instructions (avoid
-        // nonce marker), otherwise ignored
-        for (program_id, instruction) in message.program_instructions_iter().take(3) {
+        for (i, (program_id, instruction)) in message.program_instructions_iter().enumerate() {
             if compute_budget::check_id(program_id) {
-                match try_from_slice_unchecked(&instruction.data) {
-                    Ok(ComputeBudgetInstruction::RequestUnits {
-                        units,
-                        additional_fee,
-                    }) => {
-                        self.max_units = units.min(MAX_UNITS) as u64;
-                        requested_additional_fee = additional_fee as u64;
-                    }
-                    Ok(ComputeBudgetInstruction::RequestHeapFrame(bytes)) => {
-                        if !requestable_heap_size
-                            || bytes > MAX_HEAP_FRAME_BYTES
-                            || bytes < MIN_HEAP_FRAME_BYTES as u32
-                            || bytes % 1024 != 0
-                        {
-                            return Err(error);
+                // don't include request instructions in default max calc
+                num_instructions = num_instructions.saturating_sub(1);
+
+                // Compute budget instruction must be in the 1st 3 instructions (avoid
+                // nonce marker), otherwise ignored
+                if i < 3 {
+                    match try_from_slice_unchecked(&instruction.data) {
```

Comment:
OT, but should it be an error to specify any of these multiple times?

Reply:
eh, just take the last

Reply:
To be fair, it's pretty confusing when looking at the explorer and some transactions have like 10 compute budget instructions and you have to know it's the 3rd one that was applied